platforms/jvm/plugins-java-base/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java

```diff
@@ -42,38 +45,35 @@
 public interface JavaPluginExtension {
 
     /**
-     * Returns the source compatibility used for compiling Java sources.
+     * The configured source compatibility used for compiling Java sources.
+     * This property cannot be set if a {@link #getToolchain() toolchain} has been configured.
      */
-    @ToBeReplacedByLazyProperty
-    JavaVersion getSourceCompatibility();
+    @Optional
+    @ReplacesEagerProperty(adapter = JavaPluginExtensionAdapters.SourceCompatibilityAdapter.class)
+    Property<JavaVersion> getSourceCompatibility();
 
     /**
-     * Sets the source compatibility used for compiling Java sources.
-     * <p>
+     * The configured target compatibility used for compiling Java sources.
      * This property cannot be set if a {@link #getToolchain() toolchain} has been configured.
-     *
-     * @param value The value for the source compatibility
-     *
-     * @see #toolchain(Action)
      */
-    void setSourceCompatibility(JavaVersion value);
+    @Optional
+    @ReplacesEagerProperty(adapter = JavaPluginExtensionAdapters.TargetCompatibilityAdapter.class)
+    Property<JavaVersion> getTargetCompatibility();
 
     /**
-     * Returns the target compatibility used for compiling Java sources.
+     * The resulting source compatibility used for compiling Java sources with conventions applied.
+     * @since 9.0
      */
-    @ToBeReplacedByLazyProperty
-    JavaVersion getTargetCompatibility();
+    @Incubating
+    Provider<JavaVersion> getEffectiveSourceCompatibility();
```

Comment:
Ideally, there should be no need for these methods. See 
https://github.com/gradle/gradle/issues/31886